test/test_package.cpp

```diff
@@ -1436,3 +1436,103 @@ TEST(DDPackageTest, expectationValueExceptions) {
 
     EXPECT_ANY_THROW(dd->expectationValue(xGate, zeroState));
 }
+
+TEST(DDPackageTest, DDFromSingleQubitMatrix) {
+    auto row0        = dd::CVec{0, 1};
+    auto row1        = dd::CVec{2, 3};
+    auto inputMatrix = dd::CMat{row0, row1};
+
+    const dd::QubitCount nrQubits = 1;
+    auto                 dd       = std::make_unique<dd::Package<>>(nrQubits);
+    auto                 matDD    = dd->makeDDFromMatrix(inputMatrix);
+
+    auto outputMatrix = dd->getMatrix(matDD);
+
+    for (std::vector<double>::size_type i = 0; i < outputMatrix.size(); i++) {
+        for (std::vector<double>::size_type j = 0; j < outputMatrix.size(); j++) {
+            EXPECT_EQ(inputMatrix[i][j], outputMatrix[i][j]);
+        }
+    }
+}
+
+TEST(DDPackageTest, DDFromTwoQubitMatrix) {
+    auto row0        = dd::CVec{0, 1, 2, 3};
+    auto row1        = dd::CVec{4, 5, 6, 7};
+    auto row2        = dd::CVec{8, 9, 10, 11};
+    auto row3        = dd::CVec{12, 13, 14, 15};
+    auto inputMatrix = dd::CMat{row0, row1, row2, row3};
+
+    const dd::QubitCount nrQubits     = 2;
+    auto                 dd           = std::make_unique<dd::Package<>>(nrQubits);
+    auto                 matDD        = dd->makeDDFromMatrix(inputMatrix);
+    auto                 outputMatrix = dd->getMatrix(matDD);
+
+    for (std::vector<double>::size_type i = 0; i < outputMatrix.size(); i++) {
+        for (std::vector<double>::size_type j = 0; j < outputMatrix.size(); j++) {
+            EXPECT_EQ(inputMatrix[i][j], outputMatrix[i][j]);
+        }
+    }
+}
+
+TEST(DDPackageTest, DDFromThreeQubitMatrix) {
+    auto row0        = dd::CVec{0, 1, 2, 3, 4, 5, 6, 7};
+    auto row1        = dd::CVec{8, 9, 10, 11, 12, 13, 14, 15};
+    auto row2        = dd::CVec{16, 17, 18, 19, 20, 21, 22, 23};
+    auto row3        = dd::CVec{24, 25, 26, 27, 28, 29, 30, 31};
+    auto row4        = dd::CVec{32, 33, 34, 35, 36, 37, 38, 39};
+    auto row5        = dd::CVec{40, 41, 42, 43, 44, 45, 46, 47};
+    auto row6        = dd::CVec{48, 49, 50, 51, 52, 53, 54, 55};
+    auto row7        = dd::CVec{56, 57, 58, 59, 60, 61, 62, 63};
+    auto inputMatrix = dd::CMat{row0, row1, row2, row3, row4, row5, row6, row7};
+
+    const dd::QubitCount nrQubits = 3;
+    auto                 dd       = std::make_unique<dd::Package<>>(nrQubits);
+    auto                 matDD    = dd->makeDDFromMatrix(inputMatrix);
+
+    auto outputMatrix = dd->getMatrix(matDD);
+
+    for (std::vector<double>::size_type i = 0; i < outputMatrix.size(); i++) {
+        for (std::vector<double>::size_type j = 0; j < outputMatrix.size(); j++) {
+            EXPECT_EQ(inputMatrix[i][j], outputMatrix[i][j]);
+        }
+    }
```

Comment:
just as above, this can be simplified using the same techniques. Probably could also use an example of a proper quantum operation. E.g., the Toffoli would make sense here.

Reply:
Done, changed to Toffoli